common/sys/string.c: Fix `strip_nonebig5` ignoring the byte following `ff fd`

IepIweidieng opened the pull request:
This fixes #94.

According to `common/sys/big5.c`:
- Unicode code points without the corresponding Big5-UAO code point are always converted to `0xfffd` (see `u2b_table`).
- Big5-UAO code points `0xfeff`-`0xffff` are left unchanged when converted into Unicode (see `b2u_table`).

Consider the case for `strip_nonebig5`: `str` is `"\xff\xfdQ"` and `maxlen` > `3`.

When `i` is `0`, both `str[i]==255` and `251<=str[i+1] && str[i+1]<=254` hold, and the first `i++` following them skips the byte `\xff`.

`i+1<maxlen && str[i+1]` also holds, and the second `i++` skips the byte `\xfd`.

However, the `for` loop also performs `i++` after the `continue` and makes the byte `Q` ignored.

Therefore, the second `i++` is redundant and erroneous. It makes the byte following `"\xff\xfd"` ignored.

kcwu commented:
查了才發現這段 code 一開始是我寫的 XD
https://github.com/ptt/pttbbs/commit/5707bbf7d35bd2e445114451e83e6519965079ce

當初那段邏輯(251<= ... <=254)的目的其實是要濾掉 telnet 的 IAC 
http://www.tsnien.idv.tw/Internet_WebBook/chap11/11-4%20Telnet%20%E9%80%9A%E8%A8%8A%E5%8D%94%E5%AE%9A.html

後來 pttbbs 已經可以正確處理 telnet protocol, 一般輸入欄位不會再拿到 IAC sequence 了, 所以我想 `if(str[i]==255)` 那段邏輯應該可以整個砍掉了.

IepIweidieng committed:
common/sys/string.c: Fix `strip_nonebig5` ignoring the byte following `ff fd` [fixes #94]

`strip_nonebig5` was formerly `strip_iac`
    and was used to filtered out telnet protocol (RFC 854) `IAC` bytes
    to prevent the telnet clients from confusion.
See commit 7b2c37455c466e5304a0251502533cdd78ab19ab ("strip IAC char in user's nick") for details.

The following telnet command was stripped in this function:
- Data byte `0xff`:
    `IAC IAC`
    (`0xff 0xff`; but every single `0xff` is stripped)

`strip_iac` was later expanded into `strip_nonebig5`
    for filtering out non-Big5 codepoints
    to prevent the text display of client from being messed up.
The behavior of filtering out `IAC` commands was kept.
See commit 5707bbf7d35bd2e445114451e83e6519965079ce ("strip none-big5 char in user's nick") for details.

Since the commit above, the following telnet command was also stripped in this function:
- Option negotiation:
    `IAC WILL/WONT/DO/DONT <option_code>`
    (`0xff 0xfb-0xfe <option_code>`)

Later, a dedicated filter layer for handling telnet protocol was developed
    (`common/sys/telnet.c`).
It handles telnet protocol by, for example,
    adding extra `0xff` bytes for sending texts to the telnet client
    to avoid `0xff` in Big5 characters from being misinterpreted as `IAC`.

Therefore, `strip_nonebig5` now only needs to deal with data bytes,
    and the filtering of telnet commands in `strip_nonebig5` became redundant.

This caused some legal Big5 characters to be filtered out:

According to `common/sys/big5.c`:
- Unicode code points without the corresponding Big5-UAO code point
    are always converted to `0xfffd` (see `u2b_table`).
- Big5-UAO code points `0xfeff`-`0xffff` are left unchanged
    when converted into Unicode (see `b2u_table`).

Consider the case for `strip_nonebig5`:
    `str` is `"\xff\xfdQ"` and `maxlen` > `3`.

`"\xff\xfdQ"` will be misinterpreted as telnet command `IAC DO 0x51`
    and filtered out, which makes the byte `Q` ignored.

The byte following `"\xff\xfd"` will be ignored.

This commit fixes the issue
    by dropping the behavior of filtering out `IAC` commands
    for `strip_nonebig5`.

wens commented:
原本的做法看起來也不是很正確XDD

IepIweidieng commented:
As @kcwu pointed out, `strip_nonebig5` was meant to strip telnet commands (RFC 854).

Therefore `"\xff\xfdQ"` will be misinterpreted as telnet command `IAC DO 0x51` and stripped, which makes the byte `Q` ignored.

For addressing this issue, I have force pushed to update the commit to drop the code which strips the telnet commands.

kcwu commented:
LGTM

wens commented:
Could you expand the first part of your commit log a bit more?
The following is what I had written. You can see that I include more context
about why `strip_nonebig5()` strips out telnet IAC characters. Please also
include the commit references.

-------------------------------------------------------------------

strip_nonebig5() was originally strip_iac(), which filtered out telnet
protocol IAC bytes to avoid confusing the client. See commit 7b2c3745
("strip IAC char in user's nick")

It was later expanded to filter out non-Big5 codepoints to avoid messing
up displays. The original IAC-filtering code was kept. See commit
5707bbf7 ("strip none-big5 char in user's nick").

Since we now have a dedicated filter layer that handles the telnet
protocol, which includes inserting extra 0xff bytes to avoid 0xff
in Big5 characters being interpreted as IAC, we can drop this bit
of code, which has caused some legal Big5 characters being filtered
out.

IepIweidieng commented:
> Could you expand the first part of your commit log a bit more?
> The following is what I had written. You can see that I include more context
> about why `strip_nonebig5()` strips out telnet IAC characters. Please also
> include the commit references.
> 
> strip_nonebig5() was originally strip_iac(), which filtered out telnet
> protocol IAC bytes to avoid confusing the client. See commit [7b2c374](https://github.com/ptt/pttbbs/commit/7b2c37455c466e5304a0251502533cdd78ab19ab)
> ("strip IAC char in user's nick")
> 
> It was later expanded to filter out non-Big5 codepoints to avoid messing
> up displays. The original IAC-filtering code was kept. See commit
> [5707bbf](https://github.com/ptt/pttbbs/commit/5707bbf7d35bd2e445114451e83e6519965079ce) ("strip none-big5 char in user's nick").
> 
> Since we now have a dedicated filter layer that handles the telnet
> protocol, which includes inserting extra 0xff bytes to avoid 0xff
> in Big5 characters being interpreted as IAC, we can drop this bit
> of code, which has caused some legal Big5 characters being filtered
> out.

Sure!
I have force pushed to update the commit to include the information.

wens commented:
I manually picked the commit and cleaned up some issues with the commit log,
including indentation, ordering of paragraphs, and how some identifiers are
presented.

IepIweidieng commented:
After further investigation, I found out that `IAC WILL/WONT/DO/DONT <option_code>` is filtered out by `strip_nonebig5` only after 5707bbf7d35bd2e445114451e83e6519965079ce.

I have force pushed to update the commit to reflect this finding.

wens commented:
Please try not to use too much GitHub formatting in commit logs, as they are useless when viewed from a console.
Also, it would be nice if you could use your real full name and email in the future. This is standard practice in many open source projects.

IepIweidieng commented:
> Please try not to use too much GitHub formatting in commit logs, as they are useless when viewed from a console.
> Also, it would be nice if you could use your real full name and email in the future. This is standard practice in many open source projects.

Thanks for the suggestion!

wens commented:
> After further investigation, I found out that `IAC WILL/WONT/DO/DONT <option_code>` is filtered out by `strip_nonebig5` only after [5707bbf](https://github.com/ptt/pttbbs/commit/5707bbf7d35bd2e445114451e83e6519965079ce).
> 
> I have force pushed to update the commit to reflect this finding.

I updated the commit log and forced pushed the master branch.

IepIweidieng commented:
> After further investigation, I found out that `IAC WILL/WONT/DO/DONT <option_code>` is filtered out by `strip_nonebig5` only after [5707bbf](https://github.com/ptt/pttbbs/commit/5707bbf7d35bd2e445114451e83e6519965079ce).
> 
> I have force pushed to update the commit to reflect this finding.

FYI, 7b2c37455c466e5304a0251502533cdd78ab19ab made `strip_nonebig5` simply strip all `0xff` data bytes.

`common/sys/telnet.c` seems to be developed later than 7b2c37455c466e5304a0251502533cdd78ab19ab, but I am unsure whether the filtering of telnet commands in 7b2c37455c466e5304a0251502533cdd78ab19ab became an issue only after `common/sys/telnet.c` was developed.
